models/Element/Recyclebin/Item.php

```diff
@@ -329,6 +325,61 @@ protected function doRecursiveRestore(Element\ElementInterface $element)
         }
     }
 
+    /**
+     * @param ElementInterface $data
+     *
+     * @return mixed
+     */
+    public function marshalData($data)
+    {
+        //for full dump of relation fields in container types
+        $copier = new DeepCopy();
+        $copier->addTypeFilter(
+            new \DeepCopy\TypeFilter\ReplaceFilter(
+                function ($currentValue) {
+                    if ($currentValue instanceof Element\ElementInterface) {
+                        $elementType = Element\Service::getType($currentValue);
+                        $descriptor = new Model\Version\ElementDescriptor($elementType, $currentValue->getId());
+
+                        return $descriptor;
+                    }
+
+                    return $currentValue;
+                }
+            ),
+            new Model\Version\MarshalMatcher(Element\Service::getType($this->getElement()), $this->getElement()->getId())
+        );
+        $copier->addFilter(new Model\Version\SetDumpStateFilter(true), new \DeepCopy\Matcher\PropertyMatcher(Element\ElementDumpStateInterface::class, Element\ElementDumpStateInterface::DUMP_STATE_PROPERTY_NAME));
+
+        return $copier->copy($data);
+    }
+
+    /**
+     * @param $data
+     *
+     * @return mixed
+     */
+    public function unmarshalData($data)
+    {
+        $copier = new DeepCopy();
+        $copier->addTypeFilter(
+            new \DeepCopy\TypeFilter\ReplaceFilter(
+                function ($currentValue) {
+                    if ($currentValue instanceof Model\Version\ElementDescriptor) {
+                        $value = Element\Service::getElementById($currentValue->getType(), $currentValue->getId());
```

Comment:
do we need a check if destination still exists ?

Reply:
handled already!